src/main/java/org/apache/sysds/runtime/instructions/fed/UnaryMatrixFEDInstruction.java

```diff
@@ -63,17 +72,316 @@ public static UnaryMatrixFEDInstruction parseInstruction(String str) {
-			new CPOperand[]{input1}, new long[]{mo1.getFedMapping().getID()});
-		mo1.getFedMapping().execute(getTID(), true, fr1);
-		
-		//set characteristics and fed mapp
+			new CPOperand[] {input1}, new long[] {mo1.getFedMapping().getID()});
+		if(mo1.isFederated(FederationMap.FType.ROW)) {
+			FederatedRequest fr2 = new FederatedRequest(FederatedRequest.RequestType.GET_VAR, fr1.getID());
+			tmp = mo1.getFedMapping().execute(getTID(), true, fr1, fr2);
+		} else
+			tmp = mo1.getFedMapping().execute(getTID(), true, fr1);
+
+		MatrixObject out = setOutputFedMapping(ec, mo1, fr1.getID());
+
+		if(mo1.isFederated(FederationMap.FType.ROW)) {
+			aggResults(ec, mo1, tmp, out);
+		}
+	}
+
+	private void aggResults(ExecutionContext ec, MatrixObject mo1, Future<FederatedResponse>[] tmp, MatrixObject out) {
+		NewVariable tmpVar;
+		if(Arrays.asList("ucumk+", "ucum*").contains(getOpcode())) {
+			tmpVar =  getSumOrProdVariable(ec, mo1, tmp, getOpcode().equals("ucumk+"));
+			aggPartialResults(out, tmpVar);
+			ec.removeVariable(String.valueOf(tmpVar._id));
+		}
+		else if(Arrays.asList("ucummin", "ucummax").contains(getOpcode())) {
+			tmpVar = getMinOrMaxVariable(ec, mo1, tmp, getOpcode().equals("ucummin"));
+			aggPartialResults(out, tmpVar);
+			ec.removeVariable(String.valueOf(tmpVar._id));
+		} else {
+			MatrixBlock scalingValues = computeScalingValues(ec, mo1, tmp);
+			NewVariable[] ternaryVars = getTernaryVars(ec, mo1, scalingValues);
+			setScalingValues(ec, mo1, out,ternaryVars);
+			aggSumprod(out);
+		}
+	}
+
+	private void aggPartialResults(MatrixObject out, NewVariable tmpVar) {
+		CPOperand operand = new CPOperand(String.valueOf(tmpVar._id), ValueType.FP64, DataType.MATRIX);
+		modifyInstString(operand);
+
+		FederatedRequest[] fr1 = out.getFedMapping().broadcastSliced(tmpVar._mo, false);
+		FederatedRequest fr2 = FederationUtils.callInstruction(instString, output, out.getFedMapping().getID(),
+			new CPOperand[]{output, operand},
+			new long[]{out.getFedMapping().getID(), fr1[0].getID()});
+		FederatedRequest fr3 = out.getFedMapping().cleanup(getTID(), fr1[0].getID());
+		out.getFedMapping().execute(getTID(), true, fr1, fr2, fr3);
+
+		out.setFedMapping(out.getFedMapping().copyWithNewID(fr2.getID()));
+	}
+
+	private NewVariable getMinOrMaxVariable(ExecutionContext ec, MatrixObject mo1, Future<FederatedResponse>[] tmp, boolean isMin) {
+		MatrixBlock val = new MatrixBlock(1, (int) mo1.getNumColumns(), isMin ? Double.MAX_VALUE : -Double.MAX_VALUE);
+
+		int size = (int)mo1.getNumRows();
+		MatrixBlock retBlock = new MatrixBlock(size, (int)mo1.getNumColumns(), isMin ? Double.MAX_VALUE : -Double.MAX_VALUE);
+		BinaryOperator bop = InstructionUtils.parseBinaryOperator(isMin ? "min" : "max");
+
+		for(int i = 0; i < tmp.length - 1; i++)
+			try {
+				MatrixBlock curr = ((MatrixBlock) tmp[i].get().getData()[0]);
+				curr = curr.slice(curr.getNumRows()-1,curr.getNumRows()-1);
```

Comment:
The data after cumsum is as large as the input, and does allow reconstructing the input. Instead please change the implementation to want is mentioned earlier and simply specialize the code path for row and column partitioned matrices.